src/partition/repart.c

```diff
@@ -6349,6 +6353,111 @@ static int parse_partition_types(const char *p, GptPartitionType **partitions, s
         return 0;
 }
 
+#if OPENSSL_VERSION_MAJOR >= 3
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(OSSL_STORE_CTX*, OSSL_STORE_close, NULL);
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(OSSL_STORE_INFO*, OSSL_STORE_INFO_free, NULL);
+#endif
+
+static int load_openssl_provider(
+                const char *signing_engine,
+                const char *private_key_path,
+                EVP_PKEY **private_key) {
+
+        assert(signing_engine);
+        assert(private_key_path);
+        assert(private_key);
+
+#if OPENSSL_VERSION_MAJOR >= 3
+        if (!OSSL_PROVIDER_load(NULL, signing_engine))
+                return log_debug_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "Failed to load OpenSSL provider '%s': %s",
+                                signing_engine,
+                                ERR_error_string(ERR_get_error(), NULL));
+        if (!OSSL_PROVIDER_load(NULL, "default"))
+                return log_debug_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "Failed to load OpenSSL provider 'default': %s",
+                                ERR_error_string(ERR_get_error(), NULL));
+
+        _cleanup_(OSSL_STORE_closep) OSSL_STORE_CTX *store = OSSL_STORE_open(
+                        private_key_path,
+                        NULL,
+                        NULL,
+                        NULL,
+                        NULL);
+        if (!store)
+                return log_debug_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "Failed to open OpenSSL store via '%s': %s",
+                                private_key_path,
+                                ERR_error_string(ERR_get_error(), NULL));
+
+        _cleanup_(OSSL_STORE_INFO_freep) OSSL_STORE_INFO *info = OSSL_STORE_load(store);
+        if (!info && OSSL_STORE_error(store))
+                return log_debug_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "Failed to load OpenSSL store via '%s': %s",
+                                private_key_path,
+                                ERR_error_string(ERR_get_error(), NULL));
+
+        EVP_PKEY_free(*private_key);
+        *private_key = OSSL_STORE_INFO_get1_PKEY(info);
```

Comment:
so my understanding is that this doesn't actually get the private key from the token, but just a reference to it? (i mean, i hope that's the idea, otherwise the key wasn't hardware bound, after all)

Reply:
yeah it's an opaque reference that openssl knows how to use later to go talk to the provider